### DISCUSSION TITLE
Using APIViewTestCases and ViewTestCases in plugins

### DISCUSSION THREAD

**Discussion by miaow2** [netbox-community/netbox#10128] [Ideas]
Hi, NetBox team!
I am writing a plugin for NetBox and want to use **APIViewTestCases** and **ViewTestCases** in tests. But they don't work because test cases don't know that plugins namespaces are extended by `plugins:` and `plugins-api:`.  **GraphQLTestCase** does not work because function **dynamic_import** (from `utilities/utils.py`) can't import plugins graphql types.

I've made several changes and test cases started working in plugins. NetBox core tests also working. Check changes below.

In API and views test cases checks if `app_label` is in `settings.PLUGINS`

```
utilities/testing/api.py

class APITestCase(ModelTestCase):

    def _get_view_namespace(self):
        app_label = self.view_namespace or self.model._meta.app_label
        if app_label in settings.PLUGINS:
            app_label = f'plugins-api:{app_label}'
        return f'{app_label}-api'
```

```
utilities/testing/views.py

from django.conf import settings

class ModelViewTestCase(ModelTestCase):

    def _get_base_url(self):
        app_label = self.model._meta.app_label
        if app_label in settings.PLUGINS:
            app_label = f'plugins:{app_label}'
        return '{}:{}_{{}}'.format(app_label, self.model._meta.model_name)
```

In **GraphQLTestCase** I'm using `import_object` (from `extras/plugins/utils.py`) instead of `dynamic_import` for plugins.
```
utilities/api.py

from extras.plugins.utils import import_object

def get_graphql_type_for_model(model):
    """
    Return the GraphQL type class for the given model.
    """
    app_name, model_name = model._meta.label.split('.')
    # Object types for Django's auth models are in the users app
    if app_name == 'auth':
        app_name = 'users'
    if app_name in settings.PLUGINS:
        class_name = f'{app_name}.graphql.{model_name}Type'
        graphql_type = import_object(class_name)
        if graphql_type is None:
            raise GraphQLTypeNotFound(f"Could not find GraphQL type for {app_name}.{model_name}")
    else:
        class_name = f'{app_name}.graphql.types.{model_name}Type'
        try:
            graphql_type = dynamic_import(class_name)
        except AttributeError:
            raise GraphQLTypeNotFound(f"Could not find GraphQL type for {app_name}.{model_name}")

    return graphql_type
```

Tested in 3.3.0 NetBox version.

if this solution is right for you, I can create an issue and pull request. if not, feel free to close this discussion.

**Comment by abhi1693** [comment]
I think you should create an issue to discuss this further. It would be great to run test cases for my plugins this way
